base/array.jl

```diff
@@ -215,9 +215,9 @@ sizeof(a::Array) = Core.sizeof(a)
 
 function isassigned(a::Array, i::Int...)
     @inline
-    ii = (_sub2ind(size(a), i...) % UInt) - 1
-    @boundscheck ii < length(a) % UInt || return false
-    ccall(:jl_array_isassigned, Cint, (Any, UInt), a, ii) == 1
+    @boundscheck isempty(i) || checkbounds(Bool, a, i...) || return false
```

Comment:
Why the `isempty(i)`? `checkbounds(Bool, a, i...)` works with empty splats as well and correctly returns `false`, leading to `return false`. `isempty` would return `true`, passing the boundscheck through shortcircuiting and subsequently checking whether the first element is assigned for e.g. an empty vector, which doesn't make sense and would be an OOB again, right?

```
julia> Base._sub2ind(size([]))
1
```

I think just `checkbounds(Bool, a) || return false` should be sufficient.

Reply:
This was to support existing unit tests like:
```
@test isassigned(a)
```

Basically `checkbounds` returns `false` for an empty splat, but `isassigned` returns `true`.

Reply:
Seems like the latter behavior might be another bug in `isassigned`.

Reply:
After looking a little closer tonight, it's inconsistent. For example:

* `isassigned` for `Array` (the one we're talking about) returns `true` for an empty splat.
* `isassigned` for `AbstractArray` returns `false` for an empty splat.

Since they're currently inconsistent anyway, making them both return `false` sounds good. So I tried that and ran into the next issue caused by:
```
julia> checkbounds(Bool, [1])
true

julia> checkbounds(Bool, [1, 2])
false

julia> checkbounds(Bool, rand(1,1,1,1))
true
```

I think this behavior was introduced in https://github.com/JuliaLang/julia/commit/f65c403d7335acaaeb6d5ab16a6b6da679307ae6

It sounds like the intent was to preserve an ability to omit indexes for dimensions of length 1. But then arrays/vectors with one element get caught up in that. Can probably fix with a couple specializations of `checkbounds_indices`, I'll try it tomorrow.

Reply:
I think the `checkbounds` behavior for non-empty arrays is intentional. Consider these:

```
julia> [2][]
2

julia> [2;;;;;] |> typeof
Array{Int64, 5}

julia> [2;;;;;][]
2

julia> [2, 1][]
ERROR: BoundsError: attempt to access 2-element Vector{Int64} at index []
Stacktrace:
 [1] throw_boundserror(A::Vector{Int64}, I::Tuple{})
   @ Base ./abstractarray.jl:744
 [2] checkbounds
   @ ./abstractarray.jl:709 [inlined]
 [3] _getindex
   @ ./abstractarray.jl:1326 [inlined]
 [4] getindex(::Vector{Int64})
   @ Base ./abstractarray.jl:1294
 [5] top-level scope
   @ REPL[11]:1
```

Which ought to be entirely correct. So the only inconsistency I can see is for the empty case, where the current code returns `false` and yours would return true:

```
julia> ccall(:jl_array_isassigned, Cint, (Any, UInt), [], 0) == 1
true
```

Reply:
That cleared things up, thanks! Removing the `isempty` check is all that needed to be done after all then. Everything seems to be looking sane and consistent now.

Reply:
Yeah, the special casing of singleton things is sometimes a bit tricky to wrap your head around :)

Reply:
One more question for you since I think I misread this last night:

> I think the `checkbounds` behavior for non-empty arrays is a bug. Consider these:

When I run your examples through `checkbounds`, it look consistent with the results you posted:

```
julia> checkbounds(Bool, [2])
true

julia> checkbounds(Bool, [2;;;;;])
true

julia> checkbounds(Bool, [2, 1])
false
```

Is there anything else to look at there?

Reply:
No, I miswrote `checkbounds` for `isassigned` - that's why I shouldn't write these comments on a lazy holiday :sweat_smile: `checkbounds` should always agree with whether or not a `BoundsError` is thrown :)

Reply:
I probably misread it as `isassigned` last night anyway because I was up way too late. 😛 That all makes sense. Thank you!